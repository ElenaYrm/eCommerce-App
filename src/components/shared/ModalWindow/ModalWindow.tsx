import { ReactElement, useEffect } from 'react';
import styles from './modalWindow.module.scss';

interface IModalWindow {
  children: ReactElement;
  isOpen: boolean;
  onClose: () => void;
}

function ModalWindow({ children, isOpen, onClose }: IModalWindow): ReactElement | null {
  useEffect(() => {
    const handleBackdropClick = (e: MouseEvent): void => {
      if (isOpen && e.target === document.querySelector(`.${styles.root}`)) {
        onClose();
      }
    };

    if (isOpen) {
      document.body.classList.add('modal-open');
      document.addEventListener('click', handleBackdropClick);
    } else {
      document.body.classList.remove('modal-open');
      document.removeEventListener('click', handleBackdropClick);
    }

    return () => {
      document.body.classList.remove('modal-open');
      document.removeEventListener('click', handleBackdropClick);
    };
  }, [isOpen, onClose]);

  if (!isOpen) {
    return null;
  }

  return (
    <>
      {isOpen && (
        <div className={styles.root}>
          <div className={styles.root__container}>
            <div className={styles.root__closeBtn} onClick={onClose}>
              Close
            </div>
            {children}
          </div>
        </div>
      )}
    </>
  );
}

export default ModalWindow;
