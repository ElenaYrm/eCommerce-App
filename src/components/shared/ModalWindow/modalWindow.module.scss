@import '../../../styles/common';

.root {
  position: fixed;
  display: flex;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  align-items: center;
  justify-content: center;
  background-color: rgba(0, 0, 0, 0.5);
  z-index: 101;

  &__container {
    position: relative;
    display: flex;
    padding: 5rem;
    flex-direction: column;
    background-color: $light-primary;
    border-radius: 0.8rem;
  }

  &__closeBtn {
    position: absolute;
    top: 2rem;
    right: 2rem;
    cursor: pointer;
  }
}