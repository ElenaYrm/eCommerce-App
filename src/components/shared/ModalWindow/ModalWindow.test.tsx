import { render, fireEvent, screen } from '@testing-library/react';

import ModalWindow from './ModalWindow';

describe('Test ModalWindow shared component', () => {
  beforeEach(() => {
    document.body.innerHTML = '<div id="modal" />';
  });

  const onCloseMock = jest.fn();

  it('Should renders children when isOpen is true with Close button', () => {
    const { baseElement } = render(
      <ModalWindow isOpen={true} onClose={onCloseMock} isShowCloseBtn={true}>
        <div>Modal Content</div>
      </ModalWindow>,
    );

    expect(baseElement).toMatchSnapshot();
    expect(screen.queryByText('Modal Content')).toBeInTheDocument();
    expect(screen.queryByText('Close')).toBeInTheDocument();
  });

  it('Should renders children when isOpen is true without Close button', () => {
    const { baseElement } = render(
      <ModalWindow isOpen={true} onClose={onCloseMock} isShowCloseBtn={false}>
        <div>Modal Content</div>
      </ModalWindow>,
    );

    expect(baseElement).toMatchSnapshot();
    expect(screen.queryByText('Modal Content')).toBeInTheDocument();
    expect(screen.queryByText('Close')).not.toBeInTheDocument();
  });

  it('Should not render children when isOpen is false', () => {
    const { baseElement } = render(
      <ModalWindow isOpen={false} onClose={onCloseMock} isShowCloseBtn={true}>
        <div>Modal Content</div>
      </ModalWindow>,
    );

    expect(baseElement).toMatchSnapshot();
    expect(screen.queryByText('Modal Content')).toBeNull();
  });

  it('Should calls onClose when Close button is clicked', () => {
    const { baseElement } = render(
      <ModalWindow isOpen={true} onClose={onCloseMock} isShowCloseBtn={true}>
        <div>Modal Content</div>
      </ModalWindow>,
    );
    expect(baseElement).toMatchSnapshot();

    const closeButton = screen.getByText('Close');
    fireEvent.click(closeButton);
    expect(onCloseMock).toHaveBeenCalled();
  });
});
