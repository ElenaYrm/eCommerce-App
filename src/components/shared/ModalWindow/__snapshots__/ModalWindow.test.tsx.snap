// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test ModalWindow shared component Should calls onClose when Close button is clicked 1`] = `
<body
  class="stop-scroll"
>
  <div
    id="modal"
  >
    <div
      class="root"
    >
      <div
        class="root__container"
      >
        <button
          class="root__closeBtn"
          type="button"
        >
          Close
        </button>
        <div>
          Modal Content
        </div>
      </div>
    </div>
  </div>
  <div />
</body>
`;

exports[`Test ModalWindow shared component Should not render children when isOpen is false 1`] = `
<body
  class=""
>
  <div
    id="modal"
  />
  <div />
</body>
`;

exports[`Test ModalWindow shared component Should renders children when isOpen is true with Close button 1`] = `
<body
  class="stop-scroll"
>
  <div
    id="modal"
  >
    <div
      class="root"
    >
      <div
        class="root__container"
      >
        <button
          class="root__closeBtn"
          type="button"
        >
          Close
        </button>
        <div>
          Modal Content
        </div>
      </div>
    </div>
  </div>
  <div />
</body>
`;

exports[`Test ModalWindow shared component Should renders children when isOpen is true without Close button 1`] = `
<body
  class="stop-scroll"
>
  <div
    id="modal"
  >
    <div
      class="root"
    >
      <div
        class="root__container"
      >
        <div>
          Modal Content
        </div>
      </div>
    </div>
  </div>
  <div />
</body>
`;
